core/server/master/src/main/java/alluxio/master/file/meta/TrieNode.java

```diff
@@ -0,0 +1,254 @@
+ */
+@NotThreadSafe
+public final class TrieNode<T> {
+  // mChildren stores the map from T to the child TrieNode of its children
+  private Map<T, TrieNode<T>> mChildren = new HashMap<>();
+
+  // mList is valid only when mIsTerminal is true
+  private List<T> mList = null;
+
+  // mIsMountPoint is set when inserting a list of inodes
+  private boolean mIsMountPoint = false;
+
+  // mIsTerminal is true if the path is the last TrieNode of an inserted path
+  private boolean mIsTerminal = false;
+
+  /**
+   * insert a list of inodes under current TrieNode.
+   *
+   * @param inodes        inodes to be added to TrieNode
+   * @param mIsMountPoint true if inodes describe a mountPath
+   * @return the last created TrieNode based on inodes
+   */
+  public TrieNode<T> insert(List<T> inodes, boolean mIsMountPoint) {
+    TrieNode<T> current = this;
+    for (T inode : inodes) {
+      if (!current.mChildren.containsKey(inode)) {
+        current.mChildren.put(inode, new TrieNode<T>());
+      }
+      current = current.mChildren.get(inode);
+    }
+    current.mIsMountPoint = mIsMountPoint;
+    current.mIsTerminal = true;
+    if (mIsTerminal) {
+      current.mList = new ArrayList<>(mList);
+      current.mList.addAll(inodes);
+    } else {
+      current.mList = new ArrayList<>(inodes);
+    }
+    return current;
+  }
+
+  /**
+   * find the lowest matched TrieNode of given inodes.
+   *
+   * @param inodes          the target inodes
+   * @param predicate       true if this matched TrieNode must also be a mount point
+   * @param isCompleteMatch true if the TrieNode must completely match the given inodes
+   * @return null if there is no valid TrieNode, else return the lowest matched TrieNode
+   */
+  public TrieNode<T> lowestMatchedTrieNode(
+      List<T> inodes, java.util.function.Function<TrieNode, Boolean> predicate,
+      boolean isCompleteMatch) {
+    TrieNode<T> matchedPos = null;
+    TrieNode<T> current = this;
+
+    if (inodes.isEmpty() && predicate.apply(this)) {
+      return this;
+    }
+    for (int i = 0; i < inodes.size(); i++) {
+      T inode = inodes.get(i);
+      if (!current.mChildren.containsKey(inode)) {
+        if (isCompleteMatch) {
+          return null;
+        }
+        break;
+      }
+      current = current.mChildren.get(inode);
+      if (predicate.apply(current) && (!isCompleteMatch || i == inodes.size() - 1)) {
+        matchedPos = current;
+      }
+    }
+    return matchedPos;
+  }
+
+  /**
+   * find the child among current TrieNode's direct children that have key as its identifier.
+   *
+   * @param predicate the filter condition
+   * @param key the key of searched child
+   * @return not null if the valid child exists, else return null
+   */
+  public TrieNode<T> child(T key,
+                           java.util.function.Function<TrieNode<T>,
+                                              Boolean> predicate) {
+    if (isLastTrieNode()) {
+      return null;
```

Comment:
returning null is not an elegant behavior because the caller has to do null check all the time. If it's hard to make this return non-null, then the semantics(API) needs to change.